File: python/algorithm/everything_algorithm/algorithm_study/programmers/level2/2개이하로다른비트/2개이하로다른비트.py
def print_binary(n):
    return format(n, '08b')  # 8자리의 이진수 형식으로 변환

def solution(numbers):
    binary_list = [print_binary(num) for num in numbers]
    print('up : ', binary_list)

    for i in range(len(binary_list)):
        binary_str = list(binary_list[i])

        for j in range(len(binary_str) - 1, -1, -1):
            if binary_str[j] == '0':
                binary_str[j] = '1'
                break

        binary_list[i] = ''.join(binary_str)
    print(binary_list)

    return [int(binary, 2) for binary in binary_list]


print(solution([2, 7]))
# [3, 11]

# https://school.programmers.co.kr/learn/courses/30/lessons/77885

File: JavaScript/algorithm/src/algorithm/2개이하로다른비트/2개이하로다른비트.js
function solution(numbers) {
    const numberList = [];

    for (const number of numbers) {
        if (number % 2 === 0) {
            numberList.push(number + 1);
        } else {
            const binaryList = [...'0' + number.toString(2)];
            const index = binaryList.lastIndexOf('0');

            binaryList[index] = '1';
            binaryList[index + 1] = '0';
            const binary = binaryList.join('');

            numberList.push(parseInt(binary, 2));
        }
    }
    return numberList;
}

console.log(solution([2, 7]));
// # [3, 11]

// # https://school.programmers.co.kr/learn/courses/30/lessons/77885